File: 1455.check-if-a-word-occurs-as-a-prefix-of-any-word-in-a-sentence.cpp
/*
 * @lc app=leetcode id=1455 lang=cpp
 *
 * [1455] Check If a Word Occurs As a Prefix of Any Word in a Sentence
 */

#include <bits/stdc++.h>
#include "lassert/lassert.h"

using namespace std;

// @lc code=start
class Solution {
 public:
  int isPrefixOfWord(string sentence, string searchWord) {
    int ret = 1;
    int index = 0;
    int size = searchWord.size();
    for (auto c : sentence) {
      if (c == ' ') {
        ret++;
        index = 0;
      } else if (index >= 0) {
        if (c == searchWord[index]) {
          index++;
          if (index >= size) {
            return ret;
          }
        } else {
          index = -1;
        }
      }
    }
    return -1;
  }
};
// @lc code=end

int main() {
  Solution s;
  ASSERT_EQ(s.isPrefixOfWord("i love eating burger", "burg"), 4);
  ASSERT_EQ(s.isPrefixOfWord("this program is an easy problem", "pro"), 2);
  ASSERT_EQ(s.isPrefixOfWord("-1", "you"), -1);
  return 0;
}

File: 1455.check-if-a-word-occurs-as-a-prefix-of-any-word-in-a-sentence.rs
/*
 * @lc app=leetcode id=1455 lang=rust
 *
 * [1455] Check If a Word Occurs As a Prefix of Any Word in a Sentence
 */

struct Solution;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
    pub right: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

#[macro_export]
macro_rules! list_node {
    () => {
        None
    };
    ( $x:expr,$( $y:expr ),* ) => {
        {
            let mut head = Some(Box::new(ListNode::new($x)));
            let mut tail = &mut head;
            $(
                let node = Some(Box::new(ListNode::new($y)));
                tail.as_mut().unwrap().next = node;
                tail = &mut tail.as_mut().unwrap().next;
            )*
            head
        }
    };
}

// @lc code=start
impl Solution {
    pub fn is_prefix_of_word(sentence: String, search_word: String) -> i32 {
        let mut c = 1;
        let n = sentence.len();
        let mut i = 0;
        let mut b = false;
        let mut t = "".to_string();
        let sentence = sentence.chars().collect::<Vec<char>>();
        while i < n {
            if sentence[i] != ' ' {
                t += &sentence[i].to_string();
                if t == search_word {
                    b = true;
                    break;
                }
            } else {
                c += 1;
                t = "".to_string();
            }
            i += 1;
        }
        if b {
            c
        } else {
            -1
        }
    }
}
// @lc code=end

#[test]
fn test1() {
    assert_eq!(
        Solution::is_prefix_of_word("i love eating burger".to_string(), "burg".to_string()),
        4
    );
}

#[test]
fn test2() {
    assert_eq!(
        Solution::is_prefix_of_word(
            "this problem is an easy problem".to_string(),
            "pro".to_string()
        ),
        2
    );
}

#[test]
fn test3() {
    assert_eq!(
        Solution::is_prefix_of_word("i am tired".to_string(), "searchWord".to_string()),
        -1
    );
}
